docs/acs-engine-design-doc.md

```diff
@@ -0,0 +1,284 @@
+            "properties": {
+              "access": "Allow",
+              "description": "Allow kube-apiserver (tls) traffic to master",
+              "destinationAddressPrefix": "*",
+              "destinationPortRange": "443-443",
+              "direction": "Inbound",
+              "priority": 100,
+              "protocol": "Tcp",
+              "sourceAddressPrefix": "*",
+              "sourcePortRange": "*"
+            }
+          }
+        ]
+      },
+      "type": "Microsoft.Network/networkSecurityGroups"
+    }
+    {{end}}
+  ],
+  "outputs": {
+    {{range .AgentPoolProfiles}}{{template "agentoutputs.t" .}}
+    {{end}}
+    {{if not IsHostedMaster}}
+      {{template "masteroutputs.t" .}} ,
+    {{end}}
+    {{template "iaasoutputs.t" .}}
+
+  }
+}
+```
+The template generator then creates the following artifacts
+
+- ARM Templates (Deploy and Paramater JSONs). These artifacts are used by ARM to effect the actual deployment of the kubernetes clusters.
+
+- KubeConfigs. These are kubernetes config files which can be used by the user or the Kubernetes API clients to perform kubectl operations against the deployed Kubernetes cluster directly.
+
+- TLS certificates. For creating secure connections with the Azure VMs.
+
+### ARM Interface
+
+Acs-engine interfaces with Azure Resource Manager (ARM) through the Azure Go SDK. The Go SDK provides interfaces to perform functions like template deployment, validation.
+
+### Kubernetes Client API
+
+Acs-engine also performs kubernetes cluster management operations (kubectl) through the imported Kubernetes API libraries. The Client API calls are made during the scale and upgrade commands of acs-engine.
+
+
+Design challenges and proposals
+-------------------------------
+
+#### Current Challenges
+
+We find that the current implementation of templating leads to challenges in terms of code readability and maintainability.
+
+
+- There is no direct and intuitive mapping between the input apimodels and the ARM templates. The placeholder substitutions are performed at very specific areas in the template skeletons. It's hard to draw any generality from it and this makes it difficult to create the template JSONs purely through code as opposed to performing the placeholder substitutions.
+
+- This also limits the capabilities of acs-engine as far as extensibility is concerned. If we were to introduce more changes and customizations, it would potentially entail modifying the template skeleton layouts. This would just add more complexity.
+
+#### Possible Solutions
+
+**Strongly-typed schema for the ARM templates**
+
+As of now, we have no standard/formal representation of the ARM templates. They are declared as loosely typed in the Azure Go SDK.
+
+_**Pros**_
+
+- A formal representation would help us create a more direct mapping between the api model inputs and their corresponding ARM template files.
+
+- This also widens the scope for customizations in cluster definitions in the future.
```

Comment:
I would re-phrase this to be something like "This will allow us to accommodate future ARM template customization  more effectively, because we can express and maintain the variety of inter-dependent outputs natively, as first class data representations."